File: src/Systems/Swift/decorator.swift
protocol Component {
    func render() -> String
}

struct PlainMessage: Component {
    func render() -> String { "alert" }
}

class ComponentDecorator: Component {
    let inner: Component

    init(_ inner: Component) {
        self.inner = inner
    }

    func render() -> String {
        inner.render()
    }
}

final class AuditDecorator: ComponentDecorator {
    override func render() -> String { "audit(\(inner.render()))" }
}

final class EncryptDecorator: ComponentDecorator {
    override func render() -> String { "enc(\(inner.render()))" }
}

let base: Component = PlainMessage()
print("base=\(base.render())")
print("audit=\(AuditDecorator(base).render())")
print("encrypted=\(EncryptDecorator(base).render())")
print("stacked=\(AuditDecorator(EncryptDecorator(base)).render())")

File: src/Systems/Rust/decorator.rs
trait Component {
    fn render(&self) -> String;
}

struct PlainMessage;

impl Component for PlainMessage {
    fn render(&self) -> String {
        "alert".to_string()
    }
}

struct AuditDecorator {
    inner: Box<dyn Component>,
}

impl Component for AuditDecorator {
    fn render(&self) -> String {
        format!("audit({})", self.inner.render())
    }
}

struct EncryptDecorator {
    inner: Box<dyn Component>,
}

impl Component for EncryptDecorator {
    fn render(&self) -> String {
        format!("enc({})", self.inner.render())
    }
}

fn main() {
    let base = PlainMessage;
    println!("base={}", base.render());
    println!(
        "audit={}",
        AuditDecorator {
            inner: Box::new(PlainMessage),
        }
        .render()
    );
    println!(
        "encrypted={}",
        EncryptDecorator {
            inner: Box::new(PlainMessage),
        }
        .render()
    );
    println!(
        "stacked={}",
        AuditDecorator {
            inner: Box::new(EncryptDecorator {
                inner: Box::new(PlainMessage),
            }),
        }
        .render()
    );
}
